pepper-apis/src/main/java/org/broadinstitute/ddp/service/ActivityInstanceService.java

```diff
@@ -89,28 +94,65 @@ public ActivityInstanceService(ActivityInstanceDao actInstanceDao, PexInterprete
     }
 
     /**
-     * Get a form instance, translated to given language and with visibility of blocks resolved.
+     * Iterate through the instance and load activity instance summaries for each nested activity block.
      *
-     * @param handle           the jdbi handle
-     * @param userGuid         the user guid
-     * @param formInstanceGuid the form instance guid
-     * @param isoLangCode      the iso language code
-     * @return form instance, if found
-     * @throws DDPException if pex evaluation error
+     * @param handle       the database handle
+     * @param instance     the parent instance
+     * @param studyGuid    the study guid
+     * @param userGuid     the participant guid
+     * @param operatorGuid the operator guid
+     * @param isoLangCode  the preferred language iso code
      */
-    public Optional<FormInstance> getTranslatedForm(Handle handle, String userGuid, String operatorGuid, String formInstanceGuid,
-                                                    String isoLangCode, ContentStyle style) {
-        Function<ActivityInstance, FormInstance> typeChecker = (inst) -> {
-            if (ActivityType.FORMS.equals(inst.getActivityType())) {
-                return (FormInstance) inst;
-            } else {
-                LOG.warn("Expected a form instance but got type {} for guid {} lang code {}",
-                        inst.getActivityType(), formInstanceGuid, isoLangCode);
-                return null;
+    public void loadNestedInstanceSummaries(Handle handle, FormInstance instance, String studyGuid,
+                                            String userGuid, String operatorGuid, String isoLangCode) {
+        ActivityDefStore activityStore = ActivityDefStore.getInstance();
+
+        List<ActivityInstanceSummaryDto> summaryDtos = null;
+        String studyDefaultLangCode = null;
+
+        for (FormSection section : instance.getAllSections()) {
+            for (FormBlock block : section.getBlocks()) {
+                if (block.getBlockType() != BlockType.ACTIVITY) {
+                    continue;
+                }
+
+                // There might not be any nested activity blocks, so we do lazy-loading of data here.
+                if (summaryDtos == null) {
+                    summaryDtos = handle
+                            .attach(org.broadinstitute.ddp.db.dao.ActivityInstanceDao.class)
+                            .findNestedSortedInstanceSummaries(userGuid, studyGuid);
```

Comment:
Would it make sense to try to just retrieve the summaries related to the parent activity rather than all the nested summaries a user has in a study?

Reply:
Ah yes! Good point! That'll make more sense.

Reply:
Done! 98f4fa6